### PR TITLE
Out of memory issue for big file 

### PR DESCRIPTION
**Pull request by ashokmor**
Were facing out of memory issue while concurrently copying big files (600+mb) to Azure.
Refer following test case to replicate the behaviour: testBigFileUploadWithSingleThread and testConcurrentBigFileUploads of CloudBlockBlobTests.java
Before executing UTs we need to replace the dummy file "bigFile.mp4" with the real big file.
Added more inline comments to code and details to GIT commit

### CODE REVIEW

File: microsoft-azure-storage/src/com/microsoft/azure/storage/Constants.java [hunk 1]
```diff
@@ -1207,8 +1207,8 @@
     /**
      * The maximum size, in bytes, of a given stream mark operation.
      */
-    // Note if BlobConstants.MAX_SINGLE_UPLOAD_BLOB_SIZE_IN_BYTES is updated then this needs to be as well.
-    public static final int MAX_MARK_LENGTH = 256 * MB;
+    //This actually acquire the buffer into memory until file gets copied
+    public static final int MAX_MARK_LENGTH = 8 * MB;
```
Comment (jofriedm-msft) [inline microsoft-azure-storage/src/com/microsoft/azure/storage/Constants.java]: Why did you change this and MAX_SINGLE_UPLOAD_SIZE? Can you leave these values and just modify the blob request options?

File: microsoft-azure-storage/src/com/microsoft/azure/storage/blob/BlobOutputStreamInternal.java [hunk 2]
```diff
@@ -425,9 +422,9 @@ public Void call() {
 
         // Add future to set
         this.futureSet.add(this.completionService.submit(worker));
-        
+
         // Reset buffer.
-        this.outBuffer = new ByteArrayOutputStream();
+        this.outBuffer = new GettableByteArrayOutputStream();
```
Comment (jofriedm-msft) [inline microsoft-azure-storage/src/com/microsoft/azure/storage/blob/BlobOutputStreamInternal.java]: >this.outBuffer = new GettableByteArrayOutputStream(); [](start = 8, length = 53)

I think we can actually just call reset() right?

File: microsoft-azure-storage/src/com/microsoft/azure/storage/Constants.java (hunk 1 above)
Comment (ashokmor) [reply microsoft-azure-storage/src/com/microsoft/azure/storage/Constants.java]: Thank you,

Surely will do that and I think that will work too, however the default value/implementation causes issue for bug file.

File: microsoft-azure-storage/src/com/microsoft/azure/storage/blob/BlobOutputStreamInternal.java (hunk 2 above)
Comment (ashokmor) [reply microsoft-azure-storage/src/com/microsoft/azure/storage/blob/BlobOutputStreamInternal.java]: We tried that but that actually does not release the acquired bytes in memory, we have to use the sames bytes.

File: microsoft-azure-storage-test/src/com/microsoft/azure/storage/blob/CloudBlockBlobTests.java [hunk 3]
```diff
@@ -1567,6 +1579,101 @@ private void doUploadDownloadFileTest(CloudBlockBlob blob,  byte[] fileBuffer, i
...
+    @Category({ DevFabricTests.class, DevStoreTests.class, SlowTests.class })
+    public void testBigFileUploadWithSingleThread() throws IOException, StorageException, URISyntaxException {
+
+        InputStream inputStream = getBigFileToUpload();
+        final CloudBlockBlob blob = getCloudBlockBlob("bigFileTestBlob_");
+
+        this.doUploadBigFileTest(blob, inputStream, 4 * Constants.MB, 1, false, false);
+    }
+
+    //Following test case expect "bigFile.mp4" into resource folder of test module.
```
Comment (jofriedm-msft) [inline microsoft-azure-storage-test/src/com/microsoft/azure/storage/blob/CloudBlockBlobTests.java]: >bigFile.mp4 [](start = 34, length = 11)

Can you just create the file in the test? Why do you think this needs to be tested with this large of a file?

Comment (ashokmor) [reply microsoft-azure-storage-test/src/com/microsoft/azure/storage/blob/CloudBlockBlobTests.java]: We have use case of large files over cloud hence we added test case for large file. There is test case to verify large file in concurrency (5 users push file to cloud) and major memory behavior. If we prepare the large file in buffer (in memory) instead of reading it via stream (file) then that large file it self acquire all the memory.

File: microsoft-azure-storage/src/com/microsoft/azure/storage/Constants.java (hunk 1 above)
Comment (jofriedm-msft) [reply microsoft-azure-storage/src/com/microsoft/azure/storage/Constants.java]: Yes you can't modify the MAX_MARK_LENGTH but we don't want to change this value. Does lowering the max single upload size not work for you? 

---
In reply to: [149269313](https://github.com/Azure/azure-storage-java/pull/230#discussion_r149269313) [](ancestors = 149269313)

File: microsoft-azure-storage-test/src/com/microsoft/azure/storage/blob/CloudBlockBlobTests.java (hunk 3 above)
Comment (jofriedm-msft) [reply microsoft-azure-storage-test/src/com/microsoft/azure/storage/blob/CloudBlockBlobTests.java]: We actually have a framework for stressing the clients. I can follow-up to see if the scenario is being covered

---
In reply to: [149621949](https://github.com/Azure/azure-storage-java/pull/230#discussion_r149621949) [](ancestors = 149621949)

File: microsoft-azure-storage/src/com/microsoft/azure/storage/blob/BlobConstants.java [hunk 4]
```diff
@@ -178,7 +178,7 @@
     /**
      * The maximum size, in bytes, of a blob before it must be separated into blocks
      */
-    // Note if this is updated then Constants.MAX_MARK_LENGTH needs to be as well.
+    //To copy small file without buffer operation overhead
```
Comment (jofriedm-msft) [inline microsoft-azure-storage/src/com/microsoft/azure/storage/blob/BlobConstants.java]: >//To copy small file without buffer operation overhead [](start = 4, length = 54)

can you please revert the change to this file?

Comment (ashokmor) [reply microsoft-azure-storage/src/com/microsoft/azure/storage/blob/BlobConstants.java]: Updating comment back to it was?

**Comment by ashokmor** [comment]
Hello @jofriedm 

I appriciate your comment on PR, any plan to merge this to master?

We are using our updated code and it is working fine.

**Comment by jofriedm** [comment]
@mirobers to respond back

**Comment by seguler** [comment]
I believe this fix didn't resolve the issue for us - we had a different [fix ](https://github.com/Azure/azure-storage-java/commit/cc03c5f9679c8b4f8ad60d5f87a4f8574676e636)instead, and was released in v7.0 a few days ago. @mirobers

**Comment by mirobers** [comment]
Sorry for the delay! We have been busy delivering the 7.0 improvements, among which is an improvement for memory consumption as @seguler says. It significantly improves the memory profile of all uploads, with dramatic improvements for upload from file scenarios. The buffer copy in the output stream is still there so we'll need to take a fix for that part at least. I suspect that the sum total of the improvements in 7.0 are on par with what your fix provides (but we can do even better). We'll review and get back.

Quick note: I'm concerned about the change to target Java 8 as there is significant usage today on previous JRE versions. Is this change necessary?

**Comment by rickle-msft** [comment]
Because it seems like we have addressed this issue with no follow up from the customer, I am going to close this PR. Please feel free to reopen if this is still an issue.
